Tests 101

Jacky20000 opened the pull request:
## Preparations
Please check if the PR fulfills these requirements
- [ ] Readme (How to Install, Run, Use and Configure) wurde aktualisiert falls nötig
- [ ] Das Feature ist dokumentiert
- [ ] Alle nicht getter und setter Methoden sind dokumentiert
- [ ] Es wurden Unit tests für neue Features erstellt
- [ ] Alle Tests müssen erfolgreich sein
- [ ] Mindestens ein Reviewer muss den Pull Request akzeptieren

## :ticket: Description
SLADIM-101

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
feature


* **What is the current behavior?** (You can also link to an open issue here)
logs were received and saved in the database


* **What is the new behavior (if this is a feature change)?**
issue is created and reported to the backend mockup. Issue ID is assigned to the logmessage and saved in the database.


* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
The 'Logs' class was changed and a issueID was added. Function 'handleLogMessage' in logreceiver behaves differently now.


* **Other information**:
-

## :lock: Closes
Which issue(s) is (are) being closed by the PR?
-

jiaxuan1812 committed:
SLADIM-33 added issue format and graphql

Jacky20000 committed:
issue creator implementation start

SLADIM-101 logs into issue format and sending of issue (not finished)

Jacky20000 committed:
reportIssue function updated

-updated the reportIssue function so that the created Issue is sent to the MockAPI, currently with local IP address.
-dockerignore added
-Dockerfile updated
-Issue data corrected
SLADIM-147

Jacky20000 committed:
assign issue ID to logs

added assignment of the received Issue ID from the backend api to the corresponding log
-tests missing
SLADIM-148

Jacky20000 committed:
Merge branch 'SLADIM-147' into SLADIM-101

Jacky20000 committed:
Revert "Merge branch 'SLADIM-147' into SLADIM-101"

This reverts commit 557507a2012d764e7b23da9cfda862d80b49daed, reversing
changes made to b6a778f3d9b5fa1ea4a6ca512e9a54b247cb14e3.

Jacky20000 committed:
Revert "Revert "Merge branch 'SLADIM-147' into SLADIM-101""

This reverts commit 59fbd56fcb4ba4d176665fbb949098ad1d3f4a73.

Jacky20000 committed:
Revert "Revert "Revert "Merge branch 'SLADIM-147' into SLADIM-101"""

This reverts commit a4413d6e65c56452abe9c02c9c677394f8f4e38a.

Jacky20000 committed:
Merge branch 'SLADIM-101' into Tests-101

Jacky20000 committed:
Tests

added the old tests and a new test to check for issue ID

Jacky20000 committed:
Tests corrected

SLADIM-101 corrected tests and updated readme

Jacky20000 committed:
Tests again

hopefully

Jacky20000 committed:
test

Jacky20000 committed:
Revert "test"

This reverts commit 0a6c60382e5544ab864e3b5c0fb37f286b81c155.

Jacky20000 committed:
Fixing tests again

SLADIM-101 hopefully this time

Jacky20000 committed:
Corrected Doc Comments

doc comments in every file were updated and improved with return statements SLADIM-101

stiesssh commented on src/issue-creator/cb-open-issue-creator.ts:
```diff
@@ -0,0 +1,38 @@
+import { IssueFormat } from '../IssueFormat';
+/**
+ * CbOpenIssueComponent handles Timeout Logs, it extends IssueComponent to enable individual issue creation for cb open issues
+ */
```
wie ich mir den code so anschaue, erbt die klasse von 'issueReporter', nicht von 'IssueComponent.

stiesssh commented on src/issue-creator/cb-open-issue-creator.ts:
```diff
@@ -0,0 +1,38 @@
+  }
+
+   /**
+   * handles cp open logs, no specific information on how to handle cpu issues yet
+   * @param log received log in the LogMessageFormat
```
missing blank line.
it's 'cb', isn't it?
also, no need to state that log is of type LogMessageFormat, one glance at the method signature and you know the type.
```suggestion
   * handles cb open logs, no specific information on how to handle cpu issues yet
   *
   * @param log the received log
```

Jacky20000 committed:
Small update to readme

added description on how to start the database only SLADIM-101

stiesssh commented on src/issue-creator/cb-open-issue-creator.ts:
```diff
@@ -0,0 +1,38 @@
+   /**
+   * handles cp open logs, no specific information on how to handle cpu issues yet
+   * @param log received log in the LogMessageFormat
+   * @returns the issue ID received from the Back end
```
```suggestion
   * @returns the issue ID received from the backend
```

stiesssh commented on src/issue-creator/cpu-issue-creator.ts:
```diff
@@ -1,28 +1,39 @@
+import { IssueFormat } from '../IssueFormat';
 
 /**
  * CpuUtilizationIssueComponent handles CpuUtilization Logs, it extends IssueComponent to enable individual issue creation for cpu issues
```
s.o.
```suggestion
 * CpuUtilizationIssueComponent handles CpuUtilization Logs, it extends IssueReporter to enable individual issue creation for cpu issues
```

stiesssh commented:
bevor ich mit dem eigentlichen review fortfahre: 
die klassen `cb-open-issue-creator.ts`, `error-response-issue-creator.ts`, `timeout-issue-creator.ts` und `cpu-issue-creator.ts` sind bis auf eine code zeile und clone. 
refactored das bitte, z.B. in dem ihr die identischen codeabschnitte in eine methode in einer geteilten oberklasse auslagert. 

theoretisch haben die klassen mit `IssueReporter` bereits eine gemeinsame oberklasse, semantisch ist dies jedoch nicht der richtige ort, daher würde ich eine weiter klasse in die hierarchie einfügen.

dann habt ihr zwar eine klasse mehr, aber solltet ihr jemals etwas am IssueFormat oder am LogMessageFormat ändern, dann müsst ihr es nur an _einer_ stelle ändern und nicht an vieren.

stiesssh commented on src/IssueFormat.ts:
```diff
@@ -0,0 +1,17 @@
+/**
+ * This is the issue format as defined by the backend developers and
+ * messages of this type are sent to Sandro's API.
+ */
```
wer sind diese ominösen 'backend developers'? 
was ist 'Sandro's API'? 

besser wäre, wenn ihr links zu den jeweiligen repos oder die korrekten namen der jeweiligen projekte verwendent.

stiesssh commented on src/issue-creator/cb-open-issue-creator.ts:
```diff
@@ -0,0 +1,38 @@
+  async handleLog(log: LogMessageFormat) {
+    if (log.type != LogType.CB_OPEN) throw 'Wrong LogType';
+
+    const issue: IssueFormat = {
+      title: `${log.type}`,
+      body: `${log.data}`,
+      category: 'BUG',
+      componentIDs: [`${log.detector}`, `${log.source}`],
+      labels: [`${log.detector}`],
+      assignees: [`${log.detector}`],
+      locations: [`${log.source}`],
+      startDate: log.time,
+      clientMutationID: 'id1234',
+    }
+    this.id = await this.reportIssue(issue);
+    return this.id;
```
code clone. diesen teil würde ich als methode `createIssueFromLog` oder so ähnlich einen gemeinsame elternklasse auslagern, und an dieser stelle nur noch die geerbte methode aufrufen.

stiesssh commented on src/issue-creator/issue-reporter.ts:
```diff
@@ -1,25 +1,42 @@
+ * Provides the basic functionality every "detailed" IssueComponent should have
+ */
 export abstract class IssueReporter {
+  api = 'http://localhost:8080/api'; 
```
keine hardcoded urls. verwendet euer .env!

stiesssh commented on src/issue-creator/issue-reporter.ts:
```diff
@@ -1,25 +1,42 @@
-            res => console.log("Issue Report Successful!"),
-            err => console.log("Error at reporting issue")
-        )
+  /**
+   * Takes an @param issue and sends Issues to the current MockApi and receives the Issue ID
+   * Can be called in every Component that extends IssueComponent.
+   * @returns the issueID received when the the request was accepted by the server
+   */
```
wrong usage of tags!!!

```suggestion
  /**
   * sends an Issues to the current MockApi and receives the Issue ID.
   * can be called in every Component that extends IssueComponent.
   *
   * @param issue the issue to be send
   * @returns the issueID received when the the request was accepted by the server
   */
```

stiesssh commented on src/issue-creator/issue-reporter.ts:
```diff
@@ -1,25 +1,42 @@
-        )
+  /**
+   * Takes an @param issue and sends Issues to the current MockApi and receives the Issue ID
+   * Can be called in every Component that extends IssueComponent.
```
es gibt keine klasse `IssueComponent` ??

stiesssh commented on src/issue-creator/issue-reporter.ts:
```diff
@@ -1,25 +1,42 @@
-}
\ No newline at end of file
+  }
+
+
+
```
too many blank lines.

stiesssh commented on src/log-receiver/log-receiver.service.ts:
```diff
@@ -17,68 +16,78 @@ import { Logs } from 'src/schema/logs.schema';
   timeoutIssueCreator: TimeoutIssueCreatorComponent;
   cbOpenIssueCreator: CbOpenIssueCreatorComponent;
   errorResponseIssueCreator: ErrorResponseIssueCreatorComponent;
-
```
eignetlich war die leerzeile hier ganz gut. sie war der leserlichkeit zuträglich.

stiesssh commented on src/log-receiver/log-receiver.service.ts:
```diff
@@ -17,68 +16,78 @@ import { Logs } from 'src/schema/logs.schema';
   /**
    * Handling of Log messages
-   *
    * @param logMessage is the log received by the log receiver controller
+   * @returns issueID that was received from the backend
    * This calls the handleLog of the corresponding IssueCreator and passed the log message
+   * 
```
wrong order of lines. correct order is short description + blank line, more detailed description +  blank line (may be omitted if not necessary), tags.
```suggestion
   * 
   * @param logMessage is the log received by the log receiver controller
   * @returns issueID that was received from the backend
```

stiesssh commented on src/log-receiver/log-receiver.service.ts:
```diff
@@ -17,68 +16,78 @@ import { Logs } from 'src/schema/logs.schema';
   }
 
   /**
    * Handling of Log messages
```
put more info into short description to omit long description. this does not apply every where, but in this case it works, i guess.
```suggestion
   * Handles Log messages by delegating them to the respective IssueCreator
```

stiesssh commented on src/log-receiver/log-receiver.service.ts:
```diff
@@ -17,68 +16,78 @@ import { Logs } from 'src/schema/logs.schema';
   /**
-   * Writes the received log message into the database
-   *
+   * Writes the received log message into the database together with the issue ID
```
improve wording.
```suggestion
   * Writes the received log message and the issue ID into the database
```

stiesssh commented on src/log-receiver/log-receiver.service.ts:
```diff
@@ -17,68 +16,78 @@ import { Logs } from 'src/schema/logs.schema';
-   *
+   * Writes the received log message into the database together with the issue ID
+   * 
    * @param logMessage Log sent by the monitor
```
@returns tag fehlt

stiesssh commented on src/log-receiver/log-receiver.service.ts:
```diff
@@ -17,68 +16,78 @@ import { Logs } from 'src/schema/logs.schema';
   /**
    * Gets all logs from the database
-   *
-   * @returns all logs from the database
```
das @returns hat da schon hingehört.

stiesssh reviewed (changes requested):
sorry, hab zwischen durch zu früh auf 'submit review' geklickt :x 

bitte comments aus beiden reviews einarbeiten. und am besten mit meiner ersten anmerkung, bezüglich dem ändern der klassen hierarchie, anfangen.

Jacky20000 committed:
implemented the changes for the PR

changed phrasing in a couple of Doc comments, refactored the Issue creator by adding a new generic issue creator class
SLADIM-101

Jacky20000 committed:
fixing tests after newest changes

added ConfigModule to imports SLADIM-101 in both tests

stiesssh commented on src/issue-creator/cb-open-issue-creator.ts:
```diff
@@ -0,0 +1,26 @@
+/**
+ * CbOpenIssueComponent handles CB open Logs, it extends IssueCreator to enable individual issue creation for cb open issues 
+ */
+export class CbOpenIssueCreatorComponent extends IssueCreator {
+  issueCreator: IssueCreator;
+  constructor(http: HttpService, configService: ConfigService) {
+    super(http, configService);
+    this.issueCreator = new IssueCreator(http, configService);
+  }
+
+  /**
+   * handles cp open logs by creating an Issue and sending it to the API: https://github.com/ccims/ccims-backend/tree/apiMockup
+   * 
+   * @param log received log 
+   * @returns the issue ID received from the backend
+   */
+  async handleLog(log: LogMessageFormat) {
+    if (log.type != LogType.CB_OPEN) throw 'Wrong LogType';
+    return this.issueCreator.createIssueFromLog(log);
+  }
+}
```
@Jacky20000 ich glaube, du hast das konzept der vererbung nicht verstanden. ich empfehle dir, dich nochmal mit den PSE unterlagen auseinander zusetzen.

das tolle an vererbung ist schließlich, dass kinder die methoden und eigentschaften ihrere eltern erben - auch in typescript ([https://www.typescriptlang.org/docs/handbook/classes.html#inheritance](https://www.typescriptlang.org/docs/handbook/classes.html#inheritance))

zudem erachte ich es als sinnvoll, die generic IssueCreator Klasse abstract zu machen ([https://www.typescriptlang.org/docs/handbook/classes.html#abstract-classes](https://www.typescriptlang.org/docs/handbook/classes.html#abstract-classes)) und dann die methode handleLog als abstakte methode zur Klasse IssueCreator hinzuzufügen, damit alle Kindklasse diese methode implementieren müssen.

stiesssh replied:
hierüber bitte noch nachdenken und meinen vorschlag entweder einarbeiten, oder begründen warum es ein doofer vorschlag ist.

stiesssh commented on src/issue-creator/cb-open-issue-creator.ts:
```diff
@@ -0,0 +1,26 @@
+   */
+  async handleLog(log: LogMessageFormat) {
+    if (log.type != LogType.CB_OPEN) throw 'Wrong LogType';
+    return this.issueCreator.createIssueFromLog(log);
```
```suggestion
    return this.createIssueFromLog(log);
```

stiesssh commented on src/issue-creator/cb-open-issue-creator.ts:
```diff
@@ -0,0 +1,26 @@
+  issueCreator: IssueCreator;
+  constructor(http: HttpService, configService: ConfigService) {
+    super(http, configService);
+    this.issueCreator = new IssueCreator(http, configService);
```
```suggestion
```

stiesssh commented on src/issue-creator/cb-open-issue-creator.ts:
```diff
@@ -0,0 +1,26 @@
+ * CbOpenIssueComponent handles CB open Logs, it extends IssueCreator to enable individual issue creation for cb open issues 
+ */
+export class CbOpenIssueCreatorComponent extends IssueCreator {
+  issueCreator: IssueCreator;
```
```suggestion
```

stiesssh commented on src/issue-creator/cb-open-issue-creator.ts:
```diff
@@ -0,0 +1,26 @@
+  }
+
+  /**
+   * handles cp open logs by creating an Issue and sending it to the API: https://github.com/ccims/ccims-backend/tree/apiMockup
```
das sollte cb für **c**ircuit **b**reaker sein, oder?
```suggestion
   * handles cb open logs by creating an Issue and sending it to the API: https://github.com/ccims/ccims-backend/tree/apiMockup
```

Jacky20000 replied:
Ja cb steht für circuit breaker

stiesssh commented on src/issue-creator/cpu-issue-creator.ts:
```diff
@@ -1,28 +1,26 @@
+    if (log.type != LogType.CPU) throw 'Wrong LogType';
+    return this.issueCreator.createIssueFromLog(log);
+  }
+}
```
siehe cb-issue-creator. bitte die dort genannten änderung auch hier anwenden.

stiesssh commented on src/issue-creator/error-response-issue-creator.ts:
```diff
@@ -1,26 +1,26 @@
   }
-}
\ No newline at end of file
+}
```
siehe cb-issue-creator. bitte die dort genannten änderung auch hier anwenden.

stiesssh commented on src/issue-creator/issue-creator.ts:
```diff
@@ -0,0 +1,37 @@
+import { IssueFormat } from '../IssueFormat';
+import { ConfigService } from '@nestjs/config';
+/**
+ * generic IssueCreator class, it extends IssueReporter to report issues that were created issues 
+ * and it implements the IssueCreatorComponent to handle the incoming logs.
+ */
```
```suggestion
 * generic IssueCreator class. It extends IssueReporter to report issues that were created.
 * It implements the IssueCreatorComponent to handle the incoming logs.
 */
```
...zumindest, wenn du sagen will, dass 'erzeugte issues reported werden' .
außerdem machen sich satzzeichen im allgemeinen echt gut ;)

stiesssh commented on src/issue-creator/issue-reporter.ts:
```diff
@@ -1,25 +1,42 @@
-  Provides the basic funtionality every "detailed" IssueComponent should have
-*/
+/**
+ * Provides the basic functionality every "detailed" IssueComponent should have
```
hier ist die rede von den IssueCreators, oder?
```suggestion
 * Provides the basic functionality every "detailed" IssueCreator should have
```

stiesssh commented on src/issue-creator/timeout-issue-creator.ts:
```diff
@@ -1,27 +1,27 @@
-}
\ No newline at end of file
+
+}
```
siehe cb-issue-creator. bitte die dort genannten änderung auch hier anwenden.

Jacky20000 committed:
added the changes requested

SLADIM-101 changed comments and some mistakes

Jacky20000 committed:
made generic Issue Creator class abstract

IssueCreator is abstract and has the abstract method handleLog()
SLADIM-101

Jacky20000 committed:
Merge branch 'dev' into Tests-101
